File: credit.c
#include <cs50.h>
#include <stdio.h>

// Luhn's algorithm bool function prototype.
bool luhnAlg(string cardNum);

// Globally accesible variable for length of the input.
int length = 0;
int main(void)
{
    // Declaring input variable of type Long.
    long input;
    // Input prompt.
    input = get_long("number: ");
    // Copying input into variable "n", used for unmodified display of input.
    long n = input;
    // "trueInput" has the same purpose as "n" - displaying unmodified input later in code. 
    long trueInput = input;
    // "n1" and "n2" are used to single out the first two digits of the input down inside a while loop.
    long n1 = input, n2 = input;
    while (input)
    {
        n2 = n1;
        n1 = input;
        input /= 10;
    }
    // Determining length of the input via while loop and the global variable "length".
    while(n!=0)
    {
        length++;
        n = n/10;
    }
    // Convertng input stored in "trueInput" (unmodified) into a sequence of characters.
    char str[256];
    sprintf(str, "%ld", trueInput);

    // AMEX condition
    if (((n2 == 34 || n2 == 37) && length == 15) && luhnAlg(str))
    {
        printf("AMEX\n");
    }
    // MASTERCARD condition.
    else if (((n2 == 51 || n2 == 52 || n2 == 53 || n2 == 54 || n2 == 55) && length == 16) && luhnAlg(str))
    {
        printf("MASTERCARD\n");
    }
    // VISA condition
    else if (((n2 / 10) == 4 && (length == 13 || length == 16)) && luhnAlg(str))
    {
        printf("VISA\n");
    }
    // Everything else is deemed INVALID.
    else 
    {
        printf("INVALID\n");
    }

}

// Luhn's algorithm implemented in C. Output from this function goes into conditions created above.
bool luhnAlg(string cardNum)
{
    // Using global variable "length" again.
    int numberOfDigits = length;
    // Initializing final sum variable in advance.
    int sum = 0; 
    // Boolean variable used to select only every other digit of the input.
    bool isSecondDigit = false;
    // For loop, as seen, has a starting point of length -1, which is the second-to-last digit.
    for (int i = numberOfDigits - 1; i >= 0; i--)
    {
        int thisDigit = cardNum[i] - '0';

        // Multiplying every other digit by 2.
        if (isSecondDigit == true)
        {
            thisDigit = thisDigit * 2;
        }

        // Handling cases where digits sum up to a value greater than 9, aka gives 2 digits.
        sum += thisDigit / 10;
        sum += thisDigit % 10;

        // Reversing "isSecondDigit" value via NOT operator.
        isSecondDigit = !isSecondDigit;
    }
    // Returning the final sum which determines validity of the credit card number.
    return (sum % 10 == 0); 
}

File: credit.py
from cs50 import get_int


def luhn_alg(number):
    list_of_every_second = list(str(number))
    list_of_other = list(str(number))

    del list_of_every_second[-1::-2]
    multiplied_list = [int(i) * 2 for i in list_of_every_second]
    list_to_sum = []
    for k in multiplied_list:
        list_to_sum.append(k)
        if k > 9:
            list_to_sum.remove(k)
            k = int(k / 10) + (k % 10)
            list_to_sum.append(k)
    sum_of_every_second = sum(list_to_sum)
    # print(list_of_every_second)
    # print(list_to_sum)
    # print(sum_of_every_second)

    del list_of_other[-2::-2]
    # print(list_of_other)
    sum_of_other = 0
    for g in list_of_other:
        sum_of_other += int(g)
    # print(sum_of_other)
    return sum_of_every_second + sum_of_other


credit_card_input = get_int("Number: ")


if str(luhn_alg(credit_card_input))[-1] == '0' and str(credit_card_input)[:1] == '4' and (len(str(credit_card_input)) == 13 or len(str(credit_card_input)) == 16):
    print("VISA")
    # print(str(credit_card_input)[:2])

elif str(luhn_alg(credit_card_input))[-1] == '0' and len(str(credit_card_input)) == 15 and (str(credit_card_input)[:2] == '34' or str(credit_card_input)[:2] == '37'):
    print("AMEX")

elif str(luhn_alg(credit_card_input))[-1] == '0' and len(str(credit_card_input)) == 16 and (str(credit_card_input)[:2] == '51' or str(credit_card_input)[:2] == '52' or str(credit_card_input)[:2] == '53' or str(credit_card_input)[:2] == '54' or str(credit_card_input)[:2] == '55'):
    print("MASTERCARD")

else:
    print("INVALID")
    # print(luhn_alg(credit_card_input))
